hudi-client/hudi-client-common/src/main/java/org/apache/hudi/index/HoodieIndexUtils.java

```diff
@@ -251,9 +305,14 @@ private static <R> Option<HoodieRecord<R>> mergeIncomingWithExistingRecord(
       HoodieRecord<R> existing,
       Schema writeSchema,
       HoodieWriteConfig config,
-      HoodieRecordMerger recordMerger) throws IOException {
+      HoodieRecordMerger recordMerger,
+      BaseKeyGenerator keyGenerator) throws IOException {
     Schema existingSchema = HoodieAvroUtils.addMetadataFields(new Schema.Parser().parse(config.getSchema()), config.allowOperationMetadataField());
     Schema writeSchemaWithMetaFields = HoodieAvroUtils.addMetadataFields(writeSchema, config.allowOperationMetadataField());
+    if (keyGenerator != null) {
```

Comment:
we should probably avoid this.
can we check the payload config value and match it to expression payload and then route the call accordingly. 
just by looking at the code, its not obvious that keygen will be set only in case of expression payload. 
and its not future proof. what in case someone instantiates key gen at the caller side later down the line

Reply:
oh, just now looked at the lineage of this. 
I am not sure if we should think of a proper fix. 
how can we be sure that overwrite w/ latest payload works.